ocaml/xapi/xapi_xenops.ml

```diff
@@ -2033,46 +2040,60 @@ let update_vm ~__context id =
-                ( if power_state = `Suspended then
-                    let md = Xenopsd_metadata.pull ~__context id in
-                    match md.Metadata.domains with
-                    | None ->
-                        error "Suspended VM has no domain-specific metadata"
-                    | Some x ->
-                        Db.VM.set_last_booted_record ~__context ~self ~value:x ;
-                        debug "VM %s last_booted_record set to %s"
-                          (Ref.string_of self) x
-                ) ;
-                if power_state = `Halted then
-                  !trigger_xenapi_reregister ()
+
+                let update_allowed_operations () =
+                  (* Update ha_always_run before the power_state (if needed), to avoid racing
+                     with the HA monitor thread. *)
+                  let pool = Helpers.get_pool ~__context in
+                  if
+                    power_state = `Halted
+                    && not
+                         (Db.Pool.get_ha_reboot_vm_on_internal_shutdown
+                            ~__context ~self:pool
+                         )
+                  then (
+                    Db.VM.set_ha_always_run ~__context ~self ~value:false ;
+                    debug
+                      "Setting ha_always_run on vm=%s as false after shutdown"
+                      (Ref.string_of self)
+                  ) ;
+                  debug "xenopsd event: Updating VM %s power_state <- %s" id
+                    (Record_util.vm_power_state_to_string power_state) ;
+                  (* This will mark VBDs, VIFs as detached and clear resident_on
+                     if the VM has permanently shutdown.  current-operations
+                     should not be reset as there maybe a checkpoint is ongoing*)
+                  Xapi_vm_lifecycle.force_state_reset_keep_current_operations
+                    ~__context ~self ~value:power_state ;
+                  if power_state = `Running then
+                    create_guest_metrics_if_needed () ;
+                  if power_state = `Suspended || power_state = `Halted then (
+                    Xapi_network.detach_for_vm ~__context ~host:localhost
+                      ~vm:self ;
+                    Storage_access.reset ~__context ~vm:self
+                  ) ;
+                  if power_state = `Halted then
+                    Xenopsd_metadata.delete ~__context id ;
+                  ( if power_state = `Suspended then
+                      let md = Xenopsd_metadata.pull ~__context id in
+                      match md.Metadata.domains with
+                      | None ->
+                          error "Suspended VM has no domain-specific metadata"
+                      | Some x ->
+                          Db.VM.set_last_booted_record ~__context ~self ~value:x ;
+                          debug "VM %s last_booted_record set to %s"
+                            (Ref.string_of self) x
+                  ) ;
```

Comment:
Would it be simpler to move this block above the `Xapi_vm_lifecycle.force_state_reset_keep_current_operations` call (with a comment why)? I'm not too fond of the additional locking magic...

Reply:
Sounds good, will test it out

Reply:
Looks good in testing, updated accordingly.